### ISSUE TITLE
RFC: cmake, cython, and hard-coded paths/flags/tests

### ISSUE THREAD

**Issue by sarnold** [andreasvc/pyre2#19]
For several reasons I'm still on a modernize/cleanup kick, mainly for clean CI on the 3 "main" platforms, but also:

* deprecated practices/EOL python versions
* minimal/sane defaults and easy(ish) toolchain/flag configuration
* reasonable integration with OS package managers
* full cython integration with the above

I have 2 working branches that basically take opposite approaches; one uses scikit-build to abstract away almost everything (with a workaround for the non-working cython bits) and the other uses a minimal build_ext wrapper for cmake.

1) https://github.com/freepn/py-re2/tree/skbuild
2) https://github.com/freepn/py-re2/tree/cmake-minimal

The first one takes the usual cmake overrides for toolchain files, external flags, pre-init cache files, etc, but the second one actually seems to short-circuit most of the cmake magic (with a super minimal cmake file).

I'll add more here later...

**Comment by andreasvc** [comment]
Looks great!

I have no experience with either skbuild or cmake, so I wouldn't know which one is to be preferred. Do you have a preference?

BTW, I saw Windows in the list of attempted builds, is that something you're working on as well?

**Comment by sarnold** [comment]
Yes, one of the main requirements for this exercise is that it works reasonably cleanly across all 3 CI platforms, and on the Gentoo side I build/test using KEYWORDS="~amd64 ~arm ~arm64 ~x86" (same arches when I get around to making PPA packages for ubuntu).  Since I don't have windows it takes a bit of CI poking to get that working (still not working yet).

**Comment by sarnold** [comment]
Okay, due to some limitations/issues with each of the first two options above, I made a third branch:  https://github.com/freepn/py-re2/tree/pybind11 and worked through a few more "integration" issues.  Other than some workarounds for the manylinux wheels (again) in the cmake config, the main issue was pybind11 enforcing ``-fvisibility=hidden`` which ends up hiding the module init symbols so you get an import error:

```
Python 3.8.7rc1 (default, Dec 15 2020, 20:13:07) 
[GCC 10.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import re2
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: dynamic module does not define module export function (PyInit_re2)
>>> 
```
pybind11 upstream says "that should work" but I'm not sure exactly *where* the problem is so we just avoid using the pybind11 extension module for now (but still use the compile/link helpers on macos and windows).  I can make another test branch specifically for this issue, but not until after I get some other stuff done.

Other than that, everything looks pretty good in all the CI environments, including conda.  I also pushed your 0.3.2 release into the conda-forge staging repo (with the flags for ``#not win``, only py36+ and myself as recipe maintainer).

**Comment by sarnold** [comment]
Still waiting for ping reply/review:  https://github.com/conda-forge/staged-recipes/pull/13642

**Comment by sarnold** [comment]
The conda recipe I submitted was merged but sadly the conda-forge built setup does something weird (builds fine locally and in CI).  I'll submit an upstream bug for that as soon as I find the right place to do it...  Also #20 is ready for some review when you get a chance

**Comment by andreasvc** [comment]
As far as I can see all of this has now been addressed by #21
